src/Data/Form.purs

```diff
@@ -127,6 +138,126 @@ instance decodeInputSource :: (DecodeJson a) => DecodeJson (InputSource a) where
+  pure page { contents = contents }
+  where
+  evalSection :: Section Expr -> Either EvalError (Section ExprType)
+  evalSection section = do
+    contents <- traverse evalField section.contents
+    pure section { contents = contents }
+
+  evalField :: Field Expr -> Either EvalError (Field ExprType)
+  evalField field = do
+    description <- evalExpr get field.description
+    input <- evalInput field.input
+    name <- evalExpr get field.name
+    visibility <- evalExpr get field.visibility
+    pure { description, key: field.key, input, name, visibility }
+
+  evalInput :: Input Expr -> Either EvalError (Input ExprType)
+  evalInput = case _ of
+    Dropdown input -> do
+      default <- traverse (evalExpr get) input.default
+      options <- traverse (evalOption) input.options
+      valueRepresentation <- evalExpr get input.valueRepresentation
+      placeholder <- evalExpr get input.placeholder
+      required <- evalExpr get input.required
+      value <- traverse (evalExpr get) input.value
+      pure
+        ( Dropdown
+          { default
+          , index: input.index
+          , options
+          , valueRepresentation
+          , placeholder
+          , required
+          , value
+          }
+        )
+    Text input -> do
+      default <- traverse (evalExpr get) input.default
+      maxLength <- traverse (evalExpr get) input.maxLength
+      minLength <- traverse (evalExpr get) input.minLength
+      placeholder <- evalExpr get input.placeholder
+      required <- evalExpr get input.required
+      value <- traverse (evalExpr get) input.value
+      pure
+        ( Text
+          { default
+          , maxLength
+          , minLength
+          , placeholder
+          , required
+          , value
+          }
+        )
+    Toggle input -> do
+      default <- traverse (evalExpr get) input.default
+      value <- traverse (evalExpr get) input.value
+      pure (Toggle { default, value })
+
+  evalOption :: forall r. { value :: Expr | r } -> Either EvalError { value :: ExprType | r }
+  evalOption x = map (x { value = _ }) (evalExpr get x.value)
+
+keys :: Page Expr -> Map Key ExprType
+keys page = foldMap keysSection page.contents
+  where
+  keysSection :: Section Expr -> Map Key ExprType
+  keysSection section = foldMap keysField section.contents
+
+  keysField :: Field Expr -> Map Key ExprType
+  keysField field = case field.input of
+    Text { value: UserInput expr }
+      | Right value <- evalExpr (const Nothing) expr ->
+        Data.Map.singleton field.key value
+    Text { default: Just expr }
+      | Right value <- evalExpr (const Nothing) expr ->
+        Data.Map.singleton field.key value
+    Toggle { value: UserInput expr  }
+      | Right value <- evalExpr (const Nothing) expr ->
+        Data.Map.singleton field.key value
+    Toggle { default: Just expr }
+      | Right value <- evalExpr (const Nothing) expr ->
+        Data.Map.singleton field.key value
+    _ -> mempty
```

Comment:
Don't we need a `Dropdown` case here?

Reply:
Seems we could cut these in half by using the `getValue` helper

Reply:
We sure do need that case!